### ISSUE TITLE
Documenting and exposing lib/query/methods

### ISSUE THREAD

**Issue by joepie91** [tgriesser/knex#1104]
The [lib/query/methods file](https://github.com/tgriesser/knex/blob/master/lib/query/methods.js) is very useful for building Knex-based libraries (such as the ORM I'm working on right now), as it allows for chaining directly off non-Knex objects, but it's currently:

* Not documented as existing
* Not exposed directly by Knex

As it's generally bad practice to require files directly from a module (as the internal structure may change), and undocumented API surface cannot be considered to be covered by semantic versioning, it seems like a good idea to me to expose this method list directly from Knex, and document it as such. This should be a small change that would help a lot in building on top of Knex.

(Relatedly, are there any plans for a 1.0.0 release, so that semantic versioning policies can be safely applied?)

**Comment by chrisbroome** [comment]
According to #1092 , I think it's safe to say there are plans for a 1.0.0 release. As for a solid *timeline* on a 1.0 release, that's another story completely.

**Comment by tgriesser** [comment]
Yeah @joepie91 I'm wrapping up some stuff with my current job and then I'm taking a bit of time off at the beginning of 2016 to focus on knex and some side projects, expect to see a lot of activity then. 

For a 1.0 I'd really like to follow the monorepo approach babel takes and split a lot of things into very small packages all namespaced under the `@knex` scope, e.g. `@knex/methods` which would contain just what you mentioned above.

**Comment by joepie91** [comment]
I'm not sure how I feel about that. I hear a lot of complaints from people about Babel's new structure, with it being hard to understand and develop around - and in this particular case, as the goal is to obtain a list of the methods that the installed Knex version has (rather than just the newest version), I'm not sure exposing the methods as a separate module will provide that information.

**Comment by tgriesser** [comment]
@joepie91 I've worked with Babel 6 on a codebase for a few weeks, and there are still some rough edges but I believe the complaints are more around the fact that modularized configurable code transformation is a hard problem. Babel no longer provides any guarantees out of the box, so you need to install a ton of modules yourself. Knex won't go nearly this far and therefore shouldn't suffer from the same issues, and I believe it'll end up being necessary if we want to sanely manage the surface area of different dialect adapters in a semver compatible manner - planning on using https://github.com/sebmck/lerna for managing this (tool split out from Babel).

The default `require('knex')` will still under the hood be requiring all of the related modules. Most people won't want to go any further than that. For library authors, however it would be a win, since it'd make it easier to re-use just the query construction in the browser without needing to shim node related things, or be able to use the query builder without pooling, etc. 

In fact library authors might want to pick, choose, and assemble the functionality they want from knex... if you only want to use postgresql, you'd be able to depend on `@knex/builder`, `@knex/dialect-postgresql`, maybe `@knex/pool`, as the foundation for your library and go from there.

Also I actually was thinking putting methods in a package would be overkill until I saw https://github.com/jshttp/methods and actually had a decent use for it.

**Comment by joepie91** [comment]
In that sense, splitting things up into modules might be useful. I do still feel that it'd be a good idea to also expose the methods on the 'collective' `knex` module though - in some cases, an extension on Knex might want to expect it as a peerDependency, which means it has no way to require `@knex/methods` with the guarantee that it's the same version of Knex as the user installed as a peerDependency.

It might actually work best to expose the methods on a Knex instance, rather than the module. That way, an extending module can get an existing Knex query builder passed in, and reliably obtain the methods that are applicable for it.

**Comment by mercmobily** [comment]
Thanks for bringing this up, @joepie91! You're absolutely right that exposing the query methods list would be valuable for library developers building on top of Knex.

Looking at the current state of the codebase nearly a decade later, I have good news - the methods list you mentioned has evolved and is now located at `lib/query/method-constants.js`. While we never got around to the full modularization plan I mentioned back in 2016 (the `@knex` scoped packages approach), we've made progress in other ways.

Currently, the methods are being used internally when extending the QueryBuilder through `knex.QueryBuilder.extend()`, which automatically pushes new method names to this list. However, you're correct that it's not directly exposed for library authors to consume programmatically.

Your suggestion about exposing it on the Knex instance itself makes a lot of sense, especially for the peerDependency scenario you described. This would ensure library authors can always access the exact methods available in the version of Knex they're working with.

Here's what we could implement:

```javascript
// In lib/knex-builder/Knex.js
knex.QueryMethods = require('../query/method-constants');
```

And for TypeScript users:
```typescript
// In types/index.d.ts
export const QueryMethods: readonly string[];
```

This would give library developers like yourself direct access to:
```javascript
const knex = require('knex');
console.log(knex.QueryMethods); // ['with', 'withRecursive', 'select', ...]
```

It's a minimal, non-breaking change that maintains backward compatibility while providing the functionality you need. Library authors could then use this to dynamically generate wrappers, validate method calls, or build their own abstractions on top of Knex.

The implementation would just need to expose the existing `method-constants` module as `knex.QueryMethods` and add the corresponding TypeScript definitions.

**Important consideration about API stability:**

I should note that exposing this list as public API would create certain commitments on our end. Once it's part of the public API, we'd need to follow semantic versioning for any changes, which could limit our flexibility for internal refactoring. This is likely why it hasn't been exposed for the past 9 years despite requests.

Looking at our commit history, this methods list is actively maintained with regular additions (withMaterialized, clearGroup, etc.). Currently, it's only semi-public through `knex.QueryBuilder.extend()` which modifies it dynamically, but direct access to the array remains internal.

The concerns about exposing the raw array include:
- Can't change the data structure (e.g., to a Set or object) without breaking changes
- Can't add metadata about methods (like deprecation status)
- Can't reorganize how methods are categorized internally
- Risk of users mutating the array directly, potentially breaking Knex

**Alternative approach to consider:**

Instead of exposing the raw array, we could provide a read-only API that gives you the functionality while maintaining our internal flexibility:

```javascript
// Returns a frozen copy of the methods
knex.getQueryMethods() // ['with', 'withRecursive', 'select', ...]

// Check if a specific method exists
knex.hasQueryMethod('select') // true
```

This would address your use case for library authors while preserving our ability to refactor the internal architecture in the future. What do you think about this approach?

**Related issues:**
- #1536 - Chainable methods enumeration (documentation issue about the methods list)
- #2587 - Expose knex.cast function (another request to expose internal functionality)

**Comment by mercmobily** [comment]
This issue is somewhat related https://github.com/knex/knex/issues/624
